New rule: Error message is descriptive (b1e6dc)

carlosapaduarte opened the pull request:
Set of manual rules for SC 3.3.1 Error Identification

This is still missing test cases, but we're looking for approval of the concept and logic before writing them

----

## Pull Request Etiquette

### **When creating PR:**
- Make sure you requesting to **pull a branch** (right side) to the `develop` branch (left side).

### **After creating PR:**
- Add yourself (and co-authors) as "Assignees" for PR.
- Add label to indicate if it's a `Rule`, `Definition` or `Chore`.
- Close the issue that the PR resolves (and make sure the issue is referenced in the top of this comment)
- Optionally request feedback from anyone in particular by assigning them as "Reviewers".

## How to Review And Approve
- Go to the “Files changed” tab
- Here you will have the option to leave comments on different lines. 
- Once the review is completed, find the “Review changes” button in the top right, select “Approve” (if you are really confident in the rule) or "Request changes" and click “Submit review”.

jkodu commented on _rules/2045c3.md:
```diff
@@ -0,0 +1,103 @@
+---
+id: 2045c3
+name: alert Role or Live Region Identify Input Error
```
I find the casing here a bit off, please re-work.

jkodu commented on _rules/54621b.md:
```diff
@@ -0,0 +1,103 @@
+  This rule checks that the `aria-invalid` attributed is used to identify input errors
+
+accessibility_requirements: # Remove whatever is not applicable
+  wcag-technique: ARIA21: # Using Aria-Invalid to Indicate An Error Field
```
I think we do something like `aria11:state_property_processing`....

carlosapaduarte replied:
But these are tests for WCAG techniques. Why would we change this to aria?

jkodu commented on _rules/6f484a.md:
```diff
@@ -0,0 +1,101 @@
+  This rule checks that `aria-alertdialog` is used to identify input errors
+
+accessibility_requirements: # Remove whatever is not applicable
+  wcag-technique:ARIA18: # Using aria-alertdialog to Identify Errors
```
I think we do something like `aria11:state_property_processing`....

annethyme commented on _rules/2045c3.md:
```diff
@@ -0,0 +1,103 @@
+
+## Applicability
+
+This rule applies to any [document](https://www.w3.org/TR/dom/#concept-document) where the [document element](https://www.w3.org/TR/dom/#document-element) is an HTML `html` element that has at least one [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element).
```
I would suggest aligning the applicability (of all these rules) with the two other rules about form fields: 

- Form field has accessible name, https://act-rules.github.io/rules/e086e5
- Form field label is descriptive, https://act-rules.github.io/rules/cc0f0a

That would mean replacing: 

> ... that has at least one [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element).

...with something like: 

> ... that has at least one element that has one of the following [semantic roles](#semantic-role): `checkbox`, `combobox` (`select` elements), `listbox`, `menuitemcheckbox`, `menuitemradio`, `radio`, `searchbox`, `slider, spinbutton`, `switch`, `textbox`.
> 
> **Note**: The list of roles is derived by taking all the [ARIA](https://www.w3.org/TR/wai-aria-1.1/) 1.1 roles that:
> 
> - have a [semantic roles](#semantic-role) that inherits from the [abstract](https://www.w3.org/TR/wai-aria/#abstract_roles) `input` or `select` role, and
> - does not have a [required context](https://www.w3.org/TR/wai-aria/#scope) role that itself inherits from one of those roles.
> - The `option` role is not part of the list of applicable roles, because it does not meet the definition of a [User interface component](https://www.w3.org/TR/WCAG21/#dfn-user-interface-components). This means [WCAG 2.1](https://www.w3.org/TR/WCAG21/) does not require it to have an accessible name.

Another question is whether it should only apply to inputs that are [included in the accessibility tree](#included-in-the-accessibility-tree), or if it should also include those that are only [visible](#visible).

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/act-rules/act-rules.github.io?pullRequest=727) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/act-rules/act-rules.github.io?pullRequest=727) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/act-rules/act-rules.github.io?pullRequest=727) it.</sub>

jkodu commented:
@carlosapaduarte can you please remove the WIP in the heading of the PR, when you have made the changes and the PR is ready for review.

carlosapaduarte commented:
Test cases have been added to all atomic rules and the composite. Needs reviews.

audreymaniez commented on _rules/error-message-descriptive-b1e6dc.md:
```diff
@@ -0,0 +1,1942 @@
+- inherit from the [abstract](https://www.w3.org/TR/wai-aria/#abstract_roles) `input` or `select` role, and
+- do not have a [required context](https://www.w3.org/TR/wai-aria/#scope) role that itself inherits from one of those roles.
+
+## Expectation (1)
```
why is there a (1) mark?

carlosapaduarte commented on _rules/aria-alertdialog-input-error-6f484a.md:
```diff
@@ -0,0 +1,870 @@
+  - João Vicente
+---
+
+## Applicability
```
Changes made to the applicability

carlosapaduarte commented on _rules/aria-alertdialog-input-error-6f484a.md:
```diff
@@ -0,0 +1,870 @@
+
+## Expectation 
+
+After [user completion](#completed-input-field) of the target element or triggering the submission of the form, if the target element belongs to one, each target element that does not meet the instructions provided for it causes an element with `role="alertdialog"` to appear that
```
Updated

carlosapaduarte commented on _rules/aria-alertdialog-input-error-6f484a.md:
```diff
@@ -0,0 +1,870 @@
+- it is not possible to move the focus away from elements contained in the element with `role="alertdialog"` until this is dismissed; and
+- after the element with `role="alertdialog"` is dismissed the focus returns to the position where it was before the element with `role="alertdialog"` was displayed; and
+- the element with `role="alertdialog"` has an [accessible name](#accessible-name) that is not only [whitespace](#whitespace); and
+- the content of the element with `role="alertdialog"` identifies the [input error](https://www.w3.org/TR/WCAG21/#dfn-input-error).
```
Updated

carlosapaduarte commented on _rules/aria-alertdialog-input-error-6f484a.md:
```diff
@@ -0,0 +1,870 @@
+
+### Failed
+
+#### Failed Example 1
```
Updated

carlosapaduarte commented on _rules/aria-invalid-input-error-54621b.md:
```diff
@@ -0,0 +1,451 @@
+  - João Vicente
+---
+
+## Applicability
```
Updated

carlosapaduarte commented on _rules/aria-invalid-input-error-54621b.md:
```diff
@@ -0,0 +1,451 @@
+- inherit from the [abstract](https://www.w3.org/TR/wai-aria/#abstract_roles) `input` or `select` role, and
+- do not have a [required context](https://www.w3.org/TR/wai-aria/#scope) role that itself inherits from one of those roles.
+
+## Expectation 1
```
Removed

carlosapaduarte commented on _rules/aria-invalid-input-error-54621b.md:
```diff
@@ -0,0 +1,451 @@
+
+#### Failed Example 1
+
+`aria-invalid` attribute set on target element that does not have an [input error](https://www.w3.org/TR/WCAG21/#dfn-input-error).
```
Done

carlosapaduarte commented on _rules/aria-invalid-input-error-54621b.md:
````diff
@@ -0,0 +1,451 @@
+</html>
+```
+
+### Inapplicable
````
Done

carlosapaduarte commented on _rules/aria-invalid-input-error-54621b.md:
```diff
@@ -0,0 +1,451 @@
+
+### Passed
+
+#### Passed Example 1
```
Updated descriptions and added test case

carlosapaduarte commented on _rules/form-fields-not-completed-334972.md:
```diff
@@ -0,0 +1,265 @@
+  - João Vicente
+---
+
+## Applicability
```
Added

carlosapaduarte commented on _rules/form-fields-not-completed-334972.md:
```diff
@@ -0,0 +1,265 @@
+
+The content of the text message is [visible](#visible), [included in the accessibility tree](included-in-the-accessibility-tree) and indicates that the [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element) is [required](#required-input-field).
+
+## Expectation 3
```
Deleted

carlosapaduarte commented on _rules/form-fields-not-completed-334972.md:
````diff
@@ -0,0 +1,265 @@
+</form>
+```
+
+#### Passed Example 2
````
Done

carlosapaduarte commented on _rules/form-fields-not-completed-334972.md:
```diff
@@ -0,0 +1,265 @@
+
+### Failed
+
+#### Failed Example 1
```
Done

carlosapaduarte commented on _rules/form-fields-not-completed-334972.md:
````diff
@@ -0,0 +1,265 @@
+</form>
+```
+
+#### Failed Example 5
````
Done

carlosapaduarte commented on _rules/form-fields-not-completed-334972.md:
```diff
@@ -0,0 +1,265 @@
+
+### Inapplicable
+
+#### Inapplicable Example 1
```
Done

carlosapaduarte commented on _rules/aria-alertdialog-input-error-6f484a.md:
````diff
@@ -0,0 +1,870 @@
+</html>
+```
+
+### Inapplicable
````
Done

carlosapaduarte commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,287 @@
+
+## Expectation 1
+
+After [user completion](#completed-input-field) of the target element or triggering the submission of the form, if the target element does not meet the instructions provided for it, a message is presented to the user.
```
Updated

carlosapaduarte commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,287 @@
+- A tooltip displayed when the [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element) is [focused](#focusable)
+- Text content accessible through an help button or similar
+
+## Expectation 2
```
Updated

carlosapaduarte commented on _rules/invalid-form-field-value-36b590.md:
````diff
@@ -0,0 +1,287 @@
+</form>
+```
+
+#### Passed Example 2
````
Done

carlosapaduarte commented on _rules/invalid-form-field-value-36b590.md:
````diff
@@ -0,0 +1,287 @@
+<p id="age_instructions">Enter age in years</p>
+```
+
+#### Passed Example 3
````
Removed passed examples 2 and 3 (both are no longer relevant for the same reasons)

carlosapaduarte commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,287 @@
+
+### Inapplicable
+
+#### Inapplicable Example 1
```
Done

carlosapaduarte commented on _rules/live-region-input-error-2045c3.md:
```diff
@@ -0,0 +1,395 @@
+
+## Applicability
+
+This rule applies to any [document](https://www.w3.org/TR/dom/#concept-document) where the [document element](https://www.w3.org/TR/dom/#document-element) is an HTML `html` element that has at least one HTML or SVG element that has one of the following [semantic roles][semantic role]: `checkbox`, `combobox` (`select` elements), `listbox`, `menuitemcheckbox`, `menuitemradio`, `radio`, `searchbox`, ...
```
Done

carlosapaduarte commented on _rules/live-region-input-error-2045c3.md:
```diff
@@ -0,0 +1,395 @@
+- inherit from the [abstract](https://www.w3.org/TR/wai-aria/#abstract_roles) `input` or `select` role, and
+- do not have a [required context](https://www.w3.org/TR/wai-aria/#scope) role that itself inherits from one of those roles.
+
+## Expectation 1
```
Moved

carlosapaduarte commented on _rules/live-region-input-error-2045c3.md:
```diff
@@ -0,0 +1,395 @@
+
+The target element contains an HTML element with either `role=alert` or `aria-live=assertive` attribute that is present in the [document](https://www.w3.org/TR/dom/#concept-document) at the time it [completes loading](https://www.w3.org/TR/html52/dom.html#dom-documentreadystate-complete).
+
+## Expectation 2
```
Updated

carlosapaduarte commented on _rules/live-region-input-error-2045c3.md:
```diff
@@ -0,0 +1,395 @@
+- A tooltip displayed when the [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element) is [focused](#focusable)
+- Text content accessible through an help button or similar
+
+## Expectation 3
```
Updated

DagfinnRomen commented on _rules/error-message-descriptive-b1e6dc.md:
```diff
@@ -0,0 +1,1942 @@
+    - [SCR18: Providing client-side validation and alert](https://www.w3.org/WAI/WCAG21/Techniques/client-side-script/SCR18)
+    - [SCR32: Providing client-side validation and adding error text via the DOM](https://www.w3.org/WAI/WCAG21/Techniques/client-side-script/SCR32)
+
+## Test Cases
```
Align the test cases with the atomic rules.

DagfinnRomen commented on _rules/error-message-descriptive-b1e6dc.md:
```diff
@@ -0,0 +1,1019 @@
+  This rule checks that when an input error is automatically detected the error is identified and described to the user in text.
+
+accessibility_requirements:
+  wcag20:3.1.1: # Error Identification (A)
```
Editorial: 3.3.1, not 3.1.1.

jkodu commented on _rules/aria-alertdialog-input-error-6f484a.md:
```diff
@@ -0,0 +1,919 @@
+
+The rule applies to each HTML or SVG element:
+
+- the has one of the following [semantic roles](#semantic-role): `checkbox`, `combobox` (`select` elements), `listbox`, `menuitemcheckbox`, `menuitemradio`, `radio`, `searchbox`, `slider`, `spinbutton`, `switch` and `textbox`;
```
the -> that

kasperisager replied:
I'd suggest removing "(`select` elements)".

jkodu commented on _rules/aria-alertdialog-input-error-6f484a.md:
```diff
@@ -0,0 +1,919 @@
+
+## Expectation
+
+After [user completion](#completed-input-field) of the target element or triggering the submission of the form, if the target element belongs to one, each target element for which an [input error](https://www.w3.org/TR/WCAG21/#dfn-input-error) has been [automatically detected](#automatic-error-detection) causes an element with `role="alertdialog"` to appear, where the following is true:
```
can you explain `if the target element belongs to one`? I feel like this should be excluded in the applicability.

carlosapaduarte replied:
There are two possible triggers to check the expectation: 
- completion of the element
- submitting the form
But the applicability does not require the element to be in a form, that is why I've added the "if the target element belongs to one"
I'll remove the comma, which should improve the legibility

jkodu commented on pages/glossary/required-input-field.md:
```diff
@@ -0,0 +1,6 @@
+key: required-input-field
+---
+
+Any [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element) that, if left unfilled, prevents the successful submission or processing of the input.
```
Rephrased -> 

`Any input element that prevents successful submission of the form when left unfilled.`

jkodu replied:
Asides, should we also highlight the fact that, this can be any input element which as `required` attribute?

jkodu commented on pages/glossary/automatic-error-detection.md:
```diff
@@ -0,0 +1,9 @@
+key: automatic-error-detection
+---
+
+Automatic detection of [input errors](https://www.w3.org/TR/WCAG21/#dfn-input-error) is the result of:
```
This entire definition is quite vague in my opinion.

We either need links to examples of what a `specific format` is or to some best practises on how a form field should be filled.

Can we make do without this defn?

carlosapaduarte replied:
I've tried to improve this, including exemples

kasperisager commented on _rules/aria-alertdialog-input-error-6f484a.md:
```diff
@@ -0,0 +1,919 @@
+The rule applies to each HTML or SVG element:
+
+- the has one of the following [semantic roles](#semantic-role): `checkbox`, `combobox` (`select` elements), `listbox`, `menuitemcheckbox`, `menuitemradio`, `radio`, `searchbox`, `slider`, `spinbutton`, `switch` and `textbox`;
+- that may or may not belong to a [form element](https://www.w3.org/TR/html52/sec-forms.html#the-form-element);
```
Why is this needed? It's a tautology.

kasperisager commented on _rules/aria-alertdialog-input-error-6f484a.md:
```diff
@@ -0,0 +1,919 @@
+
+After [user completion](#completed-input-field) of the target element or triggering the submission of the form, if the target element belongs to one, each target element for which an [input error](https://www.w3.org/TR/WCAG21/#dfn-input-error) has been [automatically detected](#automatic-error-detection) causes an element with `role="alertdialog"` to appear, where the following is true:
+
+- it contains at least one [focusable](#focusable) element; and
```
The word "contains" is ambiguous; are we looking for a [child](https://dom.spec.whatwg.org/#concept-tree-child), a [descendant](https://dom.spec.whatwg.org/#concept-tree-descendant), a [shadow-including descendant](https://dom.spec.whatwg.org/#concept-shadow-including-descendant), or something else entirely?

kasperisager commented on _rules/aria-alertdialog-input-error-6f484a.md:
```diff
@@ -0,0 +1,919 @@
+
+- it contains at least one [focusable](#focusable) element; and
+- the focus moves to the [focusable](#focusable) element; and
+- it is not possible to move the focus away from elements contained in the element with `role="alertdialog"` until this is dismissed; and
```
The concept of being "dismissed" is not defined defined anywhere.

kasperisager commented on _rules/aria-invalid-input-error-54621b.md:
```diff
@@ -0,0 +1,468 @@
+The rule applies to each HTML or SVG element:
+
+- the has one of the following [semantic roles](#semantic-role): `checkbox`, `combobox` (`select` elements), `listbox`, `menuitemcheckbox`, `menuitemradio`, `radio`, `searchbox`, `slider`, `spinbutton`, `switch` and `textbox`;
+- that may or may not belong to a [form element](https://www.w3.org/TR/html52/sec-forms.html#the-form-element);
```
Ditto on https://github.com/act-rules/act-rules.github.io/pull/727#discussion_r332915524.

kasperisager commented on _rules/error-message-descriptive-b1e6dc.md:
```diff
@@ -0,0 +1,1019 @@
+
+## Applicability
+
+This rule applies to any [document](#https://www.w3.org/TR/dom/#concept-document) where the [document element](#https://www.w3.org/TR/dom/#document-element) is an HTML `html` element that has at least one HTML or SVG element that has one of the following [semantic roles](#semantic-role): `checkbox`, `combobox` (`select` elements), `listbox`, `menuitemcheckbox`, `menuitemradio`, `radio`, `searchbo ...
```
Ditto on https://github.com/act-rules/act-rules.github.io/pull/727#discussion_r332914863.

kasperisager commented on _rules/error-message-descriptive-b1e6dc.md:
```diff
@@ -0,0 +1,1019 @@
+
+## Expectation
+
+For each HTML or SVG element that has one of the following [semantic roles](#semantic-role): `checkbox`, `combobox` (`select` elements), `listbox`, `menuitemcheckbox`, `menuitemradio`, `radio`, `searchbox`, `slider`, `spinbutton`, `switch` and `textbox` in the test target, the outcome of at least one of the following rules is passed:
```
Ditto on https://github.com/act-rules/act-rules.github.io/pull/727#discussion_r332914863.

kasperisager commented on _rules/form-fields-not-completed-334972.md:
```diff
@@ -0,0 +1,230 @@
+
+The rule applies to each [form element](https://www.w3.org/TR/html52/sec-forms.html#the-form-element) that includes at least one [required](#required-input-field) HTML or SVG element:
+
+- that has one of the following [semantic roles][semantic role]: `checkbox`, `combobox` (`select` elements), `listbox`, `menuitemcheckbox`, `menuitemradio`, `radio`, `searchbox`, `slider`, `spinbutton`, `switch` and `textbox`.
```
Ditto on https://github.com/act-rules/act-rules.github.io/pull/727#discussion_r332914863.

kasperisager commented on _rules/form-fields-not-completed-334972.md:
```diff
@@ -0,0 +1,230 @@
+
+After triggering the submission of the target element, each [required](#required-input-field) [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element) that was not [completed](#completed-input-field) is identified by a text message.
+
+Note: A text message may identify an [input element](https://www.w3.org/TR/html52/sec-forms.html#the-input-element) in several ways, including:
```
"Note" should be emphasised.

kasperisager commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,214 @@
+
+The rule applies to each HTML or SVG element:
+
+- that has one of the following [semantic roles](#semantic-role): `checkbox`, `combobox` (`select` elements), `listbox`, `menuitemcheckbox`, `menuitemradio`, `radio`, `searchbox`, `slider`, `spinbutton`, `switch` and `textbox`;
```
Ditto on https://github.com/act-rules/act-rules.github.io/pull/727#discussion_r332914863.

kasperisager commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,214 @@
+The rule applies to each HTML or SVG element:
+
+- that has one of the following [semantic roles](#semantic-role): `checkbox`, `combobox` (`select` elements), `listbox`, `menuitemcheckbox`, `menuitemradio`, `radio`, `searchbox`, `slider`, `spinbutton`, `switch` and `textbox`;
+- that may or may not belong to a [form element](https://www.w3.org/TR/html52/sec-forms.html#the-form-element);
```
Ditto on https://github.com/act-rules/act-rules.github.io/pull/727#discussion_r332915524.

kasperisager commented on _rules/live-region-input-error-2045c3.md:
```diff
@@ -0,0 +1,341 @@
+
+This rule applies to any [document](https://www.w3.org/TR/dom/#concept-document) where the [document element](https://www.w3.org/TR/dom/#document-element) is an HTML `html` element that:
+
+- has at least one HTML or SVG element that has one of the following [semantic roles](#semantic-role): `checkbox`, `combobox` (`select` elements), `listbox`, `menuitemcheckbox`, `menuitemradio`, `radio`, `searchbox`, `slider`, `spinbutton`, `switch` and `textbox` and for which [input errors](https://www.w3.org/TR/WCAG21/#dfn-input-error) are [automatically detected](#automatic-error-detection)
```
Ditto on https://github.com/act-rules/act-rules.github.io/pull/727#discussion_r332914863.

kasperisager commented on _rules/live-region-input-error-2045c3.md:
```diff
@@ -0,0 +1,341 @@
+This rule applies to any [document](https://www.w3.org/TR/dom/#concept-document) where the [document element](https://www.w3.org/TR/dom/#document-element) is an HTML `html` element that:
+
+- has at least one HTML or SVG element that has one of the following [semantic roles](#semantic-role): `checkbox`, `combobox` (`select` elements), `listbox`, `menuitemcheckbox`, `menuitemradio`, `radio`, `searchbox`, `slider`, `spinbutton`, `switch` and `textbox` and for which [input errors](https://www.w3.org/TR/WCAG21/#dfn-input-error) are [automatically detected](#automatic-error-detection)
+- contains an HTML element with either `role=alert` or `aria-live=assertive` attributes that is present in the [document](https://www.w3.org/TR/dom/#concept-document) at the time it [completes loading](https://www.w3.org/TR/html52/dom.html#dom-documentreadystate-complete).
```
Ditto on https://github.com/act-rules/act-rules.github.io/pull/727#discussion_r332917629.

kasperisager commented on _rules/live-region-input-error-2045c3.md:
```diff
@@ -0,0 +1,341 @@
+This rule applies to any [document](https://www.w3.org/TR/dom/#concept-document) where the [document element](https://www.w3.org/TR/dom/#document-element) is an HTML `html` element that:
+
+- has at least one HTML or SVG element that has one of the following [semantic roles](#semantic-role): `checkbox`, `combobox`, `listbox`, `menuitemcheckbox`, `menuitemradio`, `radio`, `searchbox`, `slider`, `spinbutton`, `switch` and `textbox` and for which [input errors](https://www.w3.org/TR/WCAG21/#dfn-input-error) are [automatically detected](#automatic-error-detection)
+- has at least one [descendant](https://dom.spec.whatwg.org/#concept-tree-descendant) that is an HTML element with either `role=alert` or `aria-live=assertive` attributes that is present in the [document](https://www.w3.org/TR/dom/#concept-document) at the time it [completes loading](https://www.w3.org/TR/html52/dom.html#dom-documentreadystate-complete).
```
I think the formulation of "readiness" used in your other pull request is more clear: https://github.com/act-rules/act-rules.github.io/pull/960/files#diff-c66e018743316a225db096771fafa5f9R28

WilcoFiers commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,209 @@
+	function processForm() {
+		document.getElementById('error').innerText = ''
+		var age = document.getElementById('age').value
+		console.log(isNaN(age))
```
Please remove the console log.

WilcoFiers commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,209 @@
+		if (!age || isNaN(age)) {
+			document.getElementById('error').innerText = 'Age must be a number'
+		} else if (age < 1) {
+			document.getElementById('error').innerText = 'Age must be positive'
```
"Age must be at leat 1 or higher"

WilcoFiers commented on _rules/form-fields-not-completed-334972.md:
```diff
@@ -0,0 +1,227 @@
+rule_type: atomic
+description: |
+  This rule checks that text error messages identify required form fields that were not completed.
+accessibility_requirements:
```
It seems to me that this by itself pretty much is a failure of 3.3.1. Baring better definitions, I don't think this needs to be part of a composite rule.

WilcoFiers commented on _rules/aria-alertdialog-input-error-6f484a.md:
```diff
@@ -0,0 +1,915 @@
+	<head>
+		<meta charset="UTF-8" />
+		<title>Using aria-alertdialog to Identify Errors</title>
+		<script src="http://code.jquery.com/jquery.js"></script>
```
I'm not too keen to use jQuery for this. Firstly because it requires hitting the network to run these examples, but also because it create an unnecessary dependency. You can write all of this stuff in native DOM. querySelector solved a lot of problems.

carlosapaduarte commented:
Following @WilcoFiers's review (thanks for that), I've done a major overhaul of this PR. 

- I've dropped the composite rule and all the aria inspired rules that were not really checking the SC
- There are now 2 atomic rules checking different aspects of the SC: 1) required fields that have not been filled; 2) inputs that do not follow the required format or values
- The definitions have also been updated to make them more objective

The 2 atomic rules are not that different from what has been reviewed before. Hopefully, they are more objective now, but their goal remains the same.

DagfinnRomen commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,203 @@
+
+## Applicability
+
+The rule applies to each HTML [input element][] with the [type attribute][] in one of the following [states][]: "Text", "Telephone", "URL", "E-mail", "Password", "Date", "Month", "Week", "Time", "Local Date and Time", and "Number";
```
Also, the SC only applies to input errors that are automatically detected.

Add an inapplicable test case?

carlosapaduarte replied:
But that is not explicitly in the applicability of this rule (it is implicitly in the first expectation).

WilcoFiers commented on pages/glossary/automatic-error-detection.md:
```diff
@@ -0,0 +1,25 @@
+- information that is required
+- information that must follow a specific format
+- information that must be above, below or within specific values
+- information that must belong to a set of allowed values
```
```suggestion
- information that must belong to a set of allowed values
- information must be a specific value (such as in a CAPTCHA or confirming a password)
```

WilcoFiers commented on pages/glossary/automatic-error-detection.md:
```diff
@@ -0,0 +1,25 @@
+key: automatic-error-detection
+---
+
+Automatic detection of input errors is the result of automatically carried procedures to verify that the provided input meets the requirements. 
```
This doesn't actually describe what an "automatically detected input error" is. You're saying it's the result of something, but not what that thing is.

carlosapaduarte replied:
Reworded

WilcoFiers commented on pages/glossary/automatic-error-detection.md:
```diff
@@ -0,0 +1,25 @@
+The procedures can be:
+
+- Scripts executed in the client or the server;
+- Programmatically specified information that allows user agents to detect information that has not been provided or that must conform to a specific format (e.g. dates, phone numbers, etc.)
```
This is kinda vague, can you clarify?

carlosapaduarte replied:
I'm not sure what is vague in there (and I do have examples of how this can be done immediately after). Could you be more specific?

WilcoFiers commented on pages/glossary/interacted-with.md:
```diff
@@ -0,0 +1,21 @@
+Any HTML element:
+
+- with one of the following [semantic roles][semantic role]: `checkbox`, `combobox`, `listbox`, `menuitemcheckbox`, `menuitemradio`, `radio`, `searchbox`, `slider`, `spinbutton`, `switch` and `textbox`;
+- that, either through user interaction or programmatically, has had its [dirty value flag][] set to true.
```
This flag is only available for native input / select / textarea elements. That's not what the roles suggest.

carlosapaduarte replied:
This definition is no longer used. I'll remove it.

WilcoFiers commented on pages/glossary/required-field.md:
```diff
@@ -0,0 +1,21 @@
+- with one of the following [semantic roles][semantic role]: `checkbox`, `combobox`, `listbox`, `menuitemcheckbox`, `menuitemradio`, `radio`, `searchbox`, `slider`, `spinbutton`, `switch` and `textbox`;
+- that prevents the successful submission or processing of the input when empty.
+
+If the element is part of a form, having the [`required`](https://html.spec.whatwg.org/#the-required-attribute) attribute makes it a required control for form submission.
```
The required attribute can only be set on input / select / textarea elements. Your roles allow for a lot more than that.

carlosapaduarte replied:
reworded

WilcoFiers commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,203 @@
+
+## Applicability
+
+The rule applies to each HTML [input element][] with the [type attribute][] in one of the following [states][]: "Text", "Telephone", "URL", "E-mail", "Password", "Date", "Month", "Week", "Time", "Local Date and Time", and "Number";
```
I don't understand why this applicability is what it is. On the one hand, it looks like you've picked things that you know have a required format. E-mail, URL, Date, etc. Someone could programatically determine that those things exist on a page, and that submitting those with an incorrect value would result in an error. But then I don't understand why text and password are on this list. Those don't have a required format, at least not always.

On the other hand, if the intent wasn't to just limit this rule to things that have a required format, then this rule might as well include all things that inherit from ARIA input or ARIA select. That would also make this required field rule redundant.

It's also worth noting that browser won't let you set invalid values for some of these fields. You can't type letters into a type=number, type=date, type=week etc. field. Browsers enforce these things, so I'm not sure what the benefit is of testing that an error is shown for them, when it's straight up not possible to fill them out incorrectly.

If you're looking for heuristics that tell you programatically whether or not a field will be validated, but that the browser doesn't handle, the only things I know of that are like that is type=email, and type=url. Autocomplete could also be a heuristic that will tell you if there's likely a format. But those are heuristics. We'd have to make an assumption that a type=email field is actually validated. That's not an assumption I would personally be comfortable with making with our tool.

carlosapaduarte replied:
My goal was to identify form fields where a user can enter an invalid value or use an invalid format. That's why I picked the types that can be rendered by a browser as text fields (I did add a note explaining this was the criteria). 

As you said, for some of these types, browsers will prevent the user to enter invalid formats. But not all browsers will do that (Safari doesn't support type="date" for instance and falls back to type="text"). Text and password are in there for those situations where the valid format or the range of valid values is presented in the document. For example:

- Enter only your first and last names (a script can check if the user enters exactly two words)
- The password must include a lowercase letter, an uppercase letter and a number (a script will check the format)

Going with ARIA input or ARIA select would make this applicable to a lot of elements where the user cannot enter invalid values or control the format used, so in fact, they are not applicable.

Does this make sense?

carlosapaduarte replied:
You convinced me. Changing to elements with roles that inherit from ARIA input and ARIA select.

WilcoFiers commented on _rules/required-form-fields-not-completed-334972.md:
```diff
@@ -0,0 +1,243 @@
+<form>
+	<label for="text_field">Name (required)</label>
+	<input type="text" id="text_field" required />
+	<input type="button" value="Submit" />
```
This is not a submit button. The expectation calls for actually submitting the form. If you actually were to use a submit button, an error message would be shown. None of these failed cases seem valid to me.

carlosapaduarte replied:
The expectation has been reworded according to your suggestion

WilcoFiers commented on _rules/required-form-fields-not-completed-334972.md:
```diff
@@ -0,0 +1,243 @@
+
+## Applicability
+
+The rule applies to each [required][] HTML element that:
```
Browsers automatically validate form fields that have the `required` attribute. It seems to me that the only way to fail this rule is to run it in a browser that doesn't support the `required` attribute, or to put the `required` attribute on an element that doesn't support it (like `<div required role="textbox"></div>`). I don't really see how this is a useful rule, unless maybe you make this about aria-required=true?

carlosapaduarte replied:
But I'm not just using the `required` attribute in the definition of required. The goal was to have the rule also consider those instances where required fields are not marked with the `required` attribute.

WilcoFiers commented on _rules/required-form-fields-not-completed-334972.md:
```diff
@@ -0,0 +1,243 @@
+
+## Expectation 1
+
+After triggering the submission of the [form element][] that is an [ancestor][] of the target element, each target element that was not [interacted with][] is identified by a text message.
```
Same comment as for the previous rule. I think getting a form field into an error state should not be part of this rule. Just say what the requirements are for when it is in an error state.

ShadowBB commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,203 @@
+    failed: not satisfied
+    passed: satisfied
+    inapplicable: further testing needed
+input_aspects:
```
Will the fact that there are 2 techniques here make it so that if this test passes that both these techniques are satisfied? Because if so, I don't think that is correct. Per test case only one of these techniques can be satisfied, correct?

carlosapaduarte replied:
I see your point. We can have this rule pass in situations where one technique is passed and the other is inapplicable. Probably we should remove the techniques from this section. @WilcoFiers any thoughts on this?

carlosapaduarte commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,203 @@
+
+## Expectation 2
+
+The content of the message is [visible][], [included in the accessibility tree][], and identifies the [automatically detected error][].
```
Isn't stating the error message is visible and included in the accessibility tree enough?

patrickhlauke replied:
> Only nodes can be included in the accessibility tree. But someone could also make the error message available through aria-errormessage, or through aria-label(ledby) / aria-describedby. I think you should explicitly allow those options.

`aria-label` would only work for AT users, and not help users with cognitive disabilities that don't use AT (`aria-errormessage`/`aria-labelledby`/`aria-describedby` point to another element, so assuming that THAT element is visible and in the a11y tree, so that's not AT specific)

WilcoFiers commented on pages/glossary/automatic-error-detection.md:
```diff
@@ -0,0 +1,26 @@
+key: automatic-error-detection
+---
+
+Automatic detection of input errors is an automatically carried procedure to verify that the provided input meets the requirements. 
```
Another issue with this is that systems may not tell the user if an error has occurred. For example, I may build an e-shop where I let you optionally add your e-mail address. If you send me an invalid e-mail address, I might detect that but not tell you, and just to treat it as though you hadn't sent me anything. That automatic validation might even occur months later, as part of me sending my annual news letter.

I think we've got this a little backward. The problem shouldn't be about when an error occurs, but when an error is shown to the user.

carlosapaduarte replied:
Agreed. What is relevant here is that there is an error message to present to the user, not that an error occurred. I'll try to rephrase it.

carlosapaduarte commented on _rules/required-form-fields-not-completed-334972.md:
```diff
@@ -0,0 +1,233 @@
+
+The rule applies to each [required][] HTML element that:
+
+- is a [descendant][] of a [form element][]; and
```
It doesn't anymore. Previously the expectation was checking on form submission.

carlosapaduarte commented on pages/glossary/required-field.md:
```diff
@@ -0,0 +1,21 @@
+Any HTML element:
+
+- with one of the following [semantic roles][semantic role]: `checkbox`, `combobox`, `listbox`, `menuitemcheckbox`, `menuitemradio`, `radio`, `searchbox`, `slider`, `spinbutton`, `switch` and `textbox`;
+- that prevents the successful submission or processing of the input when empty.
```
Rewritten it in a way that moves the objectivity requirements to the error detection definition. Let me know if you think that the error detection definition (that has also been rephrased) is now objective.

carlosapaduarte commented:
Replaced the definition of "automatic error detection" with "error message" and updated the expectations of both rules accordingly. Please re-review @ShadowBB and @DagfinnRomen. 
@wilco, please check if you believe "error message" is not ambiguous.

patrickhlauke commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,214 @@
+
+#### Failed Example 2
+
+The error message does not identify the input field.
```
Not quite convinced the fact that the error here doesn't explicitly name the input in question is a failure. The proximity to the field that is in error should be sufficient here. Otherwise, we'd have to hard-fail any error message that doesn't explicitly identify the input field?

Perhaps a better example would be a form with 2 or more inputs, and a generic "You have errors, go and fix them" error message at the start of the form?

carlosapaduarte replied:
Indeed. I'll updated the example following your suggestion.

patrickhlauke commented:
as an aside...maybe it's just me, but i found the actual intent/understanding for 3.3.1 always a bit hard to parse. showing my ignorance, i filed this https://github.com/w3c/wcag/issues/977

DagfinnRomen commented on _rules/required-form-fields-not-completed-334972.md:
```diff
@@ -0,0 +1,238 @@
+
+## Expectation 1
+
+For each test target where the user input caused the appearance of an [error message][], the error message is [visible][] and [included in the accessibility tree][].
```
```suggestion
For each test target where the lack of user input caused the appearance of an [error message][], the error message is [visible][] and [included in the accessibility tree][].
```

WilcoFiers commented on _rules/required-form-fields-not-completed-334972.md:
```diff
@@ -0,0 +1,238 @@
+
+## Expectation 2
+
+Each [error message][] from Expectation 1 allows the identification of the test target.
```
```suggestion
Each [error message][] from Expectation 1 identifies the test target as the cause of the error.
```

WilcoFiers commented on _rules/required-form-fields-not-completed-334972.md:
```diff
@@ -0,0 +1,238 @@
+
+## Expectation 1
+
+For each test target where the lack of user input caused the appearance of an [error message][], the error message is [visible][] and [included in the accessibility tree][].
```
"lack of user input" is ambiguous. For example, someone may interpret this to mean that the user let a session expire because they went away from their machine for too long.  Failing that is not the intent of this success criterion, but the way this is written it could be misunderstood to mean this.

I also don't think the lack of input causes anything. Submitting a form causes the error. I think this could be phrased a little clearer.

carlosapaduarte replied:
I think you're backtracking a little here. We discussed this previously and because of the difficulty in objectively identifying all situations that could cause an error message the expectation focused on an error message appearing, not on what caused its appearance (e.g. submitting a form). That makes sense because the SC is concerned with the contents of the error message, not on the mechanisms (client- or served-based) that create the error message. Even if the user lets the session expire and because of that an error message is displayed, the error message should still identify the fields in an error state.

carlosapaduarte replied:
Still, I agree that we should leave causation out of the expectation and I'll replace "caused" with "resulted in". (I'll do the same for the other rule in this PR)

carlosapaduarte commented on pages/glossary/error-message.md:
```diff
@@ -0,0 +1,12 @@
+key: error-message
+---
+
+A message that is presented to the user when the user provided information does not meet the following requirements:
```
- According to SC 3.3.1, the message needs to be text
- I agree that the way the error message is presented to the user should be left out of this definition (it is already part of the expectations of rules using error message). 
- Once again, according to SC 3.3.1, user provided information is information that is entered by the user. If the browser (or device) automates some of the information entering and that results in an error message, it is my understanding that it should still be considered if the user is able to review the information entered.

I'll try to rephrase to take all this into account.

carlosapaduarte commented on pages/glossary/required-field.md:
```diff
@@ -0,0 +1,20 @@
+Any HTML element:
+
+- with one of the following [semantic roles][semantic role]: `checkbox`, `combobox`, `listbox`, `menuitemcheckbox`, `menuitemradio`, `radio`, `searchbox`, `slider`, `spinbutton`, `switch` and `textbox`;
+- that results in an [error message][] when left empty.
```
Once again, as a result of our previous discussion, we moved away from the "identify all triggers" approach.

Still, if you go for a break with a form open with elements left empty, that does not result in an error message. If there is a mechanism that automatically submits the form after a timeout or something like that and those elements are empty and it results in an error message, then those are required elements and I think this definition holds.

carlosapaduarte commented:
@WilcoFiers replied to your comments and updated expectations and error message definition

@ShadowBB @kasperisager @DagfinnRomen you guys had approved this before the changes. Can you review again? Thanks!

WilcoFiers commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,220 @@
+    - João Vicente
+---
+
+## Applicability
```
I've put several hours and a lot of thought into this. I don't think the current form works, for quite a few reasons. I'll outline a few of them as I go. But here's my counter-proposal:

```md
## Applicability

The rule applies to each HTML element that has one of the following [semantic roles][semantic role]: `checkbox`, `combobox`, `listbox`, `menuitemcheckbox`, `menuitemradio`, `radio`, `searchbox`, `slider`, `spinbutton`, `switch` or `textbox`.

**Note**: The list of applicable [semantic roles][semantic role] is derived by taking all the [ARIA 1.1][] roles that:

- inherit from the [abstract][] `input` or `select` role, and
- do not have a [required context][] role that itself inherits from one of those roles.

## Expectation 1

For each test target with a [form field error indicator][], at least one of the [form field error indicators][] allows the identification of the related test target, through [text][], or [non-text content][], or through [presentation][].

## Expectation 2

For each test target with a [form field error indicator][], at least one of the [form field error indicators][] describes the cause of the error, or how to resolve it, in [text][] that is [visible][].

## Expectation 3

For each test target with a [form field error indicator][], at least one of the [form field error indicators][] describes the cause of the error, or how to resolve it, in [text][] that is [included in the accessibility tree][] or included in the [accessible name][] or [accessible description][] of the test target.
```

carlosapaduarte commented:
@kasperisager @ShadowBB you had approved this one, so could you review again following a batch of changes from @WilcoFiers?
(check description of ea69b06a0ba8bbc24e12a812c249ede7e4986c09)

WilcoFiers commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -16,29 +16,39 @@ input_aspects:
-    - João Vicente
+	- Carlos Duarte
+	- João Vicente
+	- Wilco Fiers
```
Please fix the spacing on this.

WilcoFiers commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,220 @@
+
+### Passed
+
+#### Passed Example 1
```
This needs to be done still.

WilcoFiers commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,266 @@
+
+#### Failed Example 1
+
+The `input` element does not have a [form field error indicator][].
```
Neither does it show an error indicator. This is incorrect.

carlosapaduarte replied:
You are correct. This is in fact a passed example!

ShadowBB replied:
@WilcoFiers What exactly do you mean with "this is incorrect". Are we not allowed to use this as a passed example? This is applicable right? And it should pass this rule right? So it is a good passed example right?

WilcoFiers commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,266 @@
+	<label for="age">Age (years)</label>
+	<input type="number" id="age" />
+	<span id="error" style="position: absolute; top: -9999px; left: -9999px;"></span><br />
+	<input type="button" value="Submit" onclick="processForm()" />
```
```suggestion
	<input type="button" value="Submit" onclick="processForm()" aria-describedby="error" />
```

This will still work with display:none.

WilcoFiers commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,266 @@
+<form>
+	<label for="age">Age (years)</label>
+	<input type="number" id="age" />
+	<span id="error" style="position: absolute; top: -9999px; left: -9999px;"></span><br />
```
```suggestion
	<span id="error" style="display:none;"></span><br />
```

WilcoFiers commented on _rules/invalid-form-field-value-36b590.md:
````diff
@@ -0,0 +1,266 @@
+</form>
+```
+
+### Inapplicable
````
Include one where there's a form field without an error indicator

carlosapaduarte replied:
Having an error indicator is not part of the applicability (and it cannot be since the definition is not objective), so form fields without error indicators are still applicable.

ShadowBB commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,192 @@
+<form>
+	<label for="age">Age (years)</label>
+	<input type="number" id="age" value="0"/>
+	<span id="error">Age must be at least 1</span><br />
```
This does not describe the cause of the error. This text could also have been placed here before the user entered any value at all. This "error message" doesn't even mention the existence of an error.

ShadowBB commented on _rules/invalid-form-field-value-36b590.md:
````diff
@@ -0,0 +1,192 @@
+
+```html
+<form>
+	<h2 id="error">You must fill the name field. You must pick a color.</h2>
````
This does not describe the cause of the error. This text could also have been placed here before the user entered any value at all. This "error message" doesn't even mention the existence of an error.

ShadowBB commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,192 @@
+<form>
+	<label for="age">Age (years)</label>
+	<input type="number" id="age" value="0" />
+	<span id="error" style="display: none;">Age must be at least 1</span><br />
```
This does not describe the cause of the error.

But more importantly, the "display: none;" method of making things invisible also(!) removes it from the accessibility tree, so this would fail on multiple accounts. This example would probably be made more clear by making the text invisible but still available in the accessibility tree instead.

carlosapaduarte replied:
There is an `aria-describedby` attribute in the submit button that makes the error message available in the accessbility tree.

ShadowBB replied:
You are correct! I am not sure how I missed that.

That only leaves the "This does not describe the cause of the error" comment I guess.

carlosapaduarte replied:
I had already updated the error message. It should describe the cause of error, but let me know if you don't agree.

ShadowBB commented on _rules/invalid-form-field-value-36b590.md:
````diff
@@ -0,0 +1,192 @@
+</form>
+```
+
+#### Failed Example 5
````
Should be number "4"

ShadowBB commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,192 @@
+<form>
+	<label for="age">Age (years)</label>
+	<input type="number" id="age" value="0" />
+	<span id="error" aria-hidden="true">Age must be at least 1</span><br />
```
This does not describe the cause of the error.

ShadowBB commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,192 @@
+
+### Inapplicable
+
+#### Inapplicable Example 1
```
You might want to include an inapplicable example with a role that has a required context role that itself inherits from the ´input´ or ´select´ role.

carlosapaduarte replied:
The list of applicable elements is provided in the applicability. We already have an example without any elements from that list.

ShadowBB commented on _rules/invalid-form-field-value-36b590.md:
````diff
@@ -0,0 +1,192 @@
+</form>
+```
+
+### Failed
````
You might want to include an example that fails because although it references the label in the error message, there are multiple fields with the same label (but different fieldsets that provide further context), because it doesn't "allow for the identification of" the correct input element.

carlosapaduarte replied:
Good suggestion

carlosapaduarte commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,192 @@
+
+## Assumptions
+
+This rule assumes that the apparent cause of a [form field error indicator][] is the actual cause for the [form field error indicator][] being shown. If the actual cause is different, then this rule may generate incorrect results.
```
But the fact is that the rule is making this assumption. Do you think it's harmful to leave it here?

ShadowBB replied:
No really all that harmful. Just wrong. " If the actual cause is different, then this rule may generate incorrect results." This is categorically incorrect I think. Do you think you can come up with a scenario where this rule generates incorrect results if the cause is different?

carlosapaduarte replied:
Not really, since we are assessing the contents of the message, not the validity of the message. I'll remove the assumption.

carlosapaduarte commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,222 @@
+
+The rule applies to each HTML element that has one of the following [semantic roles][semantic role]: `checkbox`, `combobox`, `listbox`, `menuitemcheckbox`, `menuitemradio`, `radio`, `searchbox`, `slider`, `spinbutton`, `switch` or `textbox`.
+
+**Note**: The list of applicable [semantic roles][semantic role] is derived by taking all the [ARIA 1.1][] roles that:
```
We cannot remove it because the definition is in the singular form (semantic role) and in the sentence we want to use the plural form (semantic roles).

carlosapaduarte commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,222 @@
+name: Error message describes invalid form field value
+rule_type: atomic
+description: |
+  This rule checks that text descriptions are provided when the user completes a form field with invalid values or using an invalid format.
```
I do agree with you that the description is not clearly describing what the rule is now checking. I've reworded it even further.

carlosapaduarte commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,222 @@
+	<br />
+	<label for="address">Address</label>
+	<input type="text" id="address" />
+	<p>Pick a color (required)</p>
```
Added two fieldsets and associated the error message with the submit button. Is that okay?

carlosapaduarte commented:
Final call ends April 6

jkodu commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,227 @@
+
+For each test target with a [form field error indicator][], at least one of the [form field error indicators][form field error indicator] allows the identification of the related test target, through [text][], or [non-text content][], or through [presentation][].
+
+**Note**: This rule does not test [form field error indicators][form field error indicator] shown on a different page than the one of the test target.
```
```suggestion
**Note**: This rule does not test [form field error indicators][form field error indicator] shown on a different page than that of the test target.
```

jkodu commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,227 @@
+
+## Expectation 3
+
+For each test target with a [form field error indicator][], at least one of the [form field error indicators][form field error indicator] describes:
```
Both expectation 2 and 3 has the same outcome derived from a different set of values, so may be they should be just one expectation. Also the conjugation of the sentences is different with each of the expectations. Can we re-work/ rephrase this?

carlosapaduarte replied:
Isn't what you say about expectation 2 and 3, true about all expectations? :-)
I have rephrased expectation 2 to be like expectation 3.

carlosapaduarte commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,227 @@
+name: Error message describes invalid form field value
+rule_type: atomic
+description: |
+  This rule checks that text error messages provided when the user completes a form field with invalid values or using an invalid format identify the cause of the error or how to fix the error.
```
They might (it's subjective, right?). But for the description, I believe it's best to keep both so that people looking specifically for one or the other will be able to find it.

carlosapaduarte commented:
@jkodu can you check if you're happy with my changes?

Jym77 commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,241 @@
+
+## Expectation 1
+
+For each test target with a [form field error indicator][], at least one of the [form field error indicators][form field error indicator] allows the identification of the related test target, through [text][], or [non-text content][], or through [presentation][].
```
```suggestion
For each test target with a [form field error indicator][], at least one of the [form field error indicators][form field error indicator] allows the identification of the related test target, through [text][], or through [non-text content][], or through [presentation][].
```
(or remove the last "through").

Jym77 commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,241 @@
+
+#### Passed Example 1
+
+The `input` element has a [form field error indicator][] that identifies it (by referencing its label), describes the cause of the error and how to resolve it.
```
```suggestion
This `input` element has a [form field error indicator][] that identifies it (by referencing its label), describes the cause of the error and how to resolve it.
```

Jym77 commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,241 @@
+
+#### Passed Example 2
+
+The multiple `input` elements share a [form field error indicator][] that identifies the elements unfilled (by referencing their labels), describes the cause of the error and how to resolve it.
```
```suggestion
This multiple `input` elements share a [form field error indicator][] that identifies the elements unfilled (by referencing their labels), describes the cause of the error and how to resolve it.
```

Jym77 commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,241 @@
+
+#### Passed Example 3
+
+The `input` element does not have a [form field error indicator][].
```
```suggestion
This `input` element does not have a [form field error indicator][].
```

Jym77 commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,241 @@
+
+#### Failed Example 1
+
+The multiple `input` elements share a [form field error indicator][] but its message does not identify the elements that caused the error nor describes the cause of the error.
```
```suggestion
These multiple `input` elements share a [form field error indicator][] but its message does not identify the elements that caused the error nor describes the cause of the error.
```

Jym77 commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,241 @@
+
+#### Failed Example 2
+
+The `input` element has a [form field error indicator][] but its message does not describe the cause of the error.
```
```suggestion
This `input` element has a [form field error indicator][] but its message does not describe the cause of the error.
```

Jym77 commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,241 @@
+
+#### Failed Example 3
+
+The `input` element has a [form field error indicator][] that identifies it (by referencing its label) and describes the cause of the error but the message is not [visible][].
```
```suggestion
This `input` element has a [form field error indicator][] that identifies it (by referencing its label) and describes the cause of the error but the message is not [visible][].
```

Jym77 commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,241 @@
+
+#### Failed Example 4
+
+The `input` element has a [form field error indicator][] that identifies it (by referencing its label) and describes the cause of the error but the message is not [included in the accessibility tree][].
```
```suggestion
This `input` element has a [form field error indicator][] that identifies it (by referencing its label) and describes the cause of the error but the message is not [included in the accessibility tree][].
```

Jym77 commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,241 @@
+
+#### Failed Example 5
+
+The multiple `input` elements share a [form field error indicator][]. The message describes the cause of the error but does not allow to identify the elements that caused the error because the same label is used in different `fieldset`s.
```
```suggestion
These multiple `input` elements share a [form field error indicator][]. The message describes the cause of the error but does not allow to identify the elements that caused the error because the same label is used in different `fieldset`s.
```

Jym77 commented on _rules/invalid-form-field-value-36b590.md:
```diff
@@ -0,0 +1,241 @@
+
+#### Inapplicable Example 1
+
+No elements with the required [semantic roles][semantic role].
```
```suggestion
There is no elements with any of the required [semantic roles][semantic role].
```
